sockeye/train.py

```diff
@@ -132,278 +139,460 @@ def main():
+    :param training_state_dir: Directory where the training state is stored.
+    :return: The training model.
+    """
+    training_model = training.TrainingModel(config=model_config,
+                                            context=context,
+                                            train_iter=train_iter,
+                                            fused=args.use_fused_rnn,
+                                            bucketing=not args.no_bucketing,
+                                            lr_scheduler=lr_scheduler_instance)
+
+    # We may consider loading the params in TrainingModule, for consistency
+    # with the training state saving
+    if resume_training:
+        logger.info("Found partial training in directory %s. Resuming from saved state.", training_state_dir)
+        training_model.load_params_from_file(os.path.join(training_state_dir, C.TRAINING_STATE_PARAMS_NAME))
+    elif args.params:
+        logger.info("Training will initialize from parameters loaded from '%s'", args.params)
+        training_model.load_params_from_file(args.params)
+
+    return training_model
+
+
+def define_optimizer(args, lr_scheduler_instance) -> Tuple[str, Dict]:
+    """
+    Defines the optimizer to use and its parameters
+
+    :param args: Arguments as returned by argparse.
+    :param lr_scheduler: The learning rate scheduler.
+    :return: The optimizer type and its parameters.
+    """
+    optimizer = args.optimizer
+    optimizer_params = {'wd': args.weight_decay,
+                        "learning_rate": args.initial_learning_rate}
+    if lr_scheduler_instance is not None:
+        optimizer_params["lr_scheduler"] = lr_scheduler_instance
+    clip_gradient = none_if_negative(args.clip_gradient)
+    if clip_gradient is not None:
+        optimizer_params["clip_gradient"] = clip_gradient
+    if args.momentum is not None:
+        optimizer_params["momentum"] = args.momentum
+    if args.normalize_loss:
+        # When normalize_loss is turned on we normalize by the number of non-PAD symbols in a batch which implicitly
+        # already contains the number of sentences and therefore we need to disable rescale_grad.
+        optimizer_params["rescale_grad"] = 1.0
+    else:
+        # Making MXNet module API's default scaling factor explicit
+        optimizer_params["rescale_grad"] = 1.0 / args.batch_size
+    logger.info("Optimizer: %s", optimizer)
+    logger.info("Optimizer Parameters: %s", optimizer_params)
+
+    return optimizer, optimizer_params
+
 
-            # write vocabularies
-            vocab.vocab_to_json(vocab_source, os.path.join(output_folder, C.VOCAB_SRC_NAME) + C.JSON_SUFFIX)
-            vocab.vocab_to_json(vocab_target, os.path.join(output_folder, C.VOCAB_TRG_NAME) + C.JSON_SUFFIX)
+def main():
+    params = argparse.ArgumentParser(description='CLI to train sockeye sequence-to-sequence models.')
+    arguments.add_train_cli_args(params)
+    args = params.parse_args()
+
+    seedRNGs(args)
```

Comment:
seedRNGs could probably be moved to utils.py. It might be useful for other CLIs at some point.